src/macaron/slsa_analyzer/git_service/__init__.py

```diff
@@ -1,13 +1,115 @@
-# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2022 - 2023, Oracle and/or its affiliates. All rights reserved.
 # Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
 
 """The git_service package contains the supported git services for Macaron."""
 
-from .base_git_service import BaseGitService
+import os
+from configparser import ConfigParser
+
+from macaron.errors import ConfigurationError
+
+from .base_git_service import BaseGitService, NoneGitService
 from .bitbucket import BitBucket
 from .github import GitHub
 from .gitlab import GitLab
 
-# The list of supported git services. The order of the list determines the order
-# in which each git service is checked against the target repository.
-GIT_SERVICES: list[BaseGitService] = [GitHub(), GitLab(), BitBucket()]
+__all__ = [
+    "BaseGitService",
+    "BitBucket",
+    "GitHub",
+    "GitLab",
+    "NoneGitService",
+]
+
+
+def load_git_services(config: ConfigParser) -> list[BaseGitService]:
+    """Load git services from a ``.ini`` configuration.
+
+    Parameters
+    ----------
+    config : ConfigParser
+        The ``.ini`` configuration.
+
+    Returns
+    -------
+    list[BaseGitService]
+        The list of git services.
+
+    Raises
+    ------
+    ConfigurationError
+        If the configuration for git services is invalid.
+    """
+    git_service_section_names = [
+        section_name for section_name in config.sections() if section_name.startswith("git_service")
+    ]
+
+    git_services: list[BaseGitService] = []
+
+    for section_name in git_service_section_names:
+        git_service_section = config[section_name]
+
+        # Domain is compulsory in every ``git_service.*`` section.
+        domain = git_service_section.get("domain")
+        if not domain:
+            raise ConfigurationError(
+                f'The "domain" key is missing in section [{section_name}] of the .ini configuration file.'
+            )
+
+        if section_name.startswith("git_service.github"):
```

Comment:
What's the reason to move loading defaults here and using `if`-`else` to handle each service separately?  The intention to have a [`load_defaults()` function](https://github.com/oracle/macaron/pull/316/files#diff-833baa746fec95ad48a6fdccb975226a06a0668dd98a214103275861689de193L23) to be implemented per Git service was to avoid such an `if-else` implementation.

Reply:
The way I do it is: I create one object of the `GitLab` class for each GitLab service (public or private). Each `GitLab` object currently does not know which section of the config it should look into (either `[git_service.gitlab.public]` or `[git_service.gitlab.private]`) to populate itself with the domain and access token.

Reply:
You can set the values like [here](https://github.com/oracle/macaron/blob/2f658bde1f185a0d265ee05f8f31eb02bfd7fd96/src/macaron/slsa_analyzer/ci_service/github_actions.py#L85) as part of the `load_defaults` function. You can also have two subclasses for the public and private versions of  the `GitLab` for handling the tokens differently.

Reply:
There is another reason for this `load_git_services` function, which is configuration validation (you can see what I do [here](https://github.com/oracle/macaron/blob/6754bb2e9b026a13f97f070f53f9c63ad33f8942/src/macaron/__main__.py#L140)). The way I think about it, this validation should be done before any git service object is created, not the other way around.

Reply:
The validation should happen earlier [here](https://github.com/oracle/macaron/blob/2f658bde1f185a0d265ee05f8f31eb02bfd7fd96/src/macaron/__main__.py#L278). I will address that in a separate PR as it is not directly related to this feature.

Reply:
> You can also have two subclasses for the public and private versions of GitLab for handling the tokens differently.
I don't like the idea of introducing another level of abstraction here.
In addition, doing it like this means we must always "speculatively" create 1 public GitLab service object and 1 private GitLab service object. In reality, neither can be necessary at all because a user does not enable either of them in their config.
Also by having this `load_defaults`, we are somehow introducing different states to the git service objects, i.e. after being created with the constructor, the git service object is not yet ready to be used as it also has to call this `load_defaults` method to get ready. 
Another issue is: what if the configuration is invalid, e.g. there's no domain in the section? Do we throw the object away?

Reply:
> The validation of the ini file happens [here](https://github.com/oracle/macaron/blob/c1b5435961e2b61f5377f0cc8a441c0fff28aa55/src/macaron/__main__.py#L340) already as part of the load_defaults function.

By validation, I mean also checking the semantic of the configuration file, not just checking whether we can load it in or not.

Reply:
Previously, the `load_defaults` method, although not really doing anything, getting called over and over again [here](https://github.com/oracle/macaron/blob/102853cfeaf318391b8d66d2e6302b2ce4e5c359/src/macaron/slsa_analyzer/analyzer.py#L580). The call stack here is `get_git_service <- _prepare_repo <- run_single <- run`. `run_single` is called multiple times, not just for the main repo but also for its dependency.
This is not great. Optimally, loading config should only be done once when the object is created, or right after the fact.
I feel like the fact that we are introducing states into the object is part of the reason why things like this can happen. We either need to be extra careful, or we don't have the "created but config not loaded" state in the first place.

Reply:
Two points:

1. If the domain provided for a Git service in the ini file is not valid, Macaron run should not proceed and exit early.
2. The current design in general assumes certain objects exist, so that we can iterate through them without constantly checking if they are `None` or not. Please follow this design.

Reply:
I think the reason why we need to do it differently this time is because:
- The initialization of the base git services depend on the combination of values we have in the `.ini` config file especially in the case of GitLab private and public. Therefore, we need to read the config values first before initializing the BaseGitService instances accordingly. If we were doing it using the original method, I thing it would be more complicated given the time limitation (e.g we need to check if we are not creating 2 GitLab instances public and private at the same time as it doesn't make sense, how would we return if during the analysis, the GitLab instance when try to load the default values would recognise that the environmental variable is not set correctly). 
- I think the reason why I put `GIT_SERVICES: list[BaseGitService] = [GitHub(), GitLab(), BitBucket()]` originally is because the original Git services did not load the values from defaults.ini so it worked at that time because we đidn't need to prepare those class instances with any values from users.
- I just recognize that because `get_git_service` method is called again and again [here](https://github.com/oracle/macaron/blob/102853cfeaf318391b8d66d2e6302b2ce4e5c359/src/macaron/slsa_analyzer/analyzer.py#L580), we could potentially not read the same original values (i.e if the env var changed during the analysis , etc.)

Reply:
@tromai 
>we need to check if we are not creating 2 GitLab instances public and private at the same time as it doesn't make sense

We can have two public and private instances at the same time. Assume a case where a main target is private but the dependencies are using the public GitLab service.

Reply:
> 1. If the domain provided for a Git service is not valid, Macaron run should not proceed and exit early.
I wonder if you mean that the `load_defaults` method in each BaseGitService should detect if there are errors in the `.ini` config? Hmmm I think there is a potential issue with this approach. For example, if we are creating both PublicGitLab and PrivateGitLab instances beforehand. During the analysis we run `load_defaults` for PublicGitLab first, at this stage of execution, if the config for the PublicGitLab is invalid, should we exit or we only exit after checking if PrivateGitLab config exist/valid because exiting. This could introduce more complexity to the implementation I think. :thinking: 

> 2. The current design in general assumes certain objects exist, so that we can iterate through them without constantly checking if they are None or not. Please follow this design.
Could we store the list of BaseGitService returned in the Analyzer and iterate through them instead ? :thinking:

Reply:
> We can have two public and private instances at the same time. Assume a case where a main target is private but the dependencies are using the public GitLab service.

Ah I think I get this part now. I think we could either:
- Have a class for PublicGitLab and PrivateGitLab or
- Have one GitLab class but two instances created for each of them (I personally think this would be better as the only difference is the token value and the domain value (I assume cloning, etc for them is the same), it's not necessary for us to create two sub class given that they are not that much different).

Reply:
>I think the reason why I put GIT_SERVICES: list[BaseGitService] = [GitHub(), GitLab(), BitBucket()] originally is because the original Git services did not load the values from defaults.ini so it worked at that time because we đidn't need to prepare those class instances with any values from users.

The values were loaded [here](https://github.com/oracle/macaron/blob/c1b5435961e2b61f5377f0cc8a441c0fff28aa55/src/macaron/slsa_analyzer/analyzer.py#L580) though were loaded again and again before cloning a repo (which was not necessary).

Reply:
> The values were loaded [here](https://github.com/oracle/macaron/blob/c1b5435961e2b61f5377f0cc8a441c0fff28aa55/src/macaron/slsa_analyzer/analyzer.py#L580) though were loaded again and again before cloning a repo (which was not necessary).

I can revert the `GIT_SERVICES` variable and load the config into it. This ensures that we do not repeatedly call this method over again.

Reply:
I think this conversation can be resolved as we have changed the implementation s.

Reply:
I think we agreed to leave this to a later refactor PR.